### PR TITLE
feat: Added more field in consultation form

### LINKED ISSUE #859: More Fields to be added in Consultation Api
Part of [#2218](https://github.com/coronasafe/care_fe/issues/2218)
2 collapsable sections would be added to the consultation form **Health Section** and **Medical History** extra fields need to be added as the information given in the issue [#2218](https://github.com/coronasafe/care_fe/issues/2218)

### PR DESCRIPTION
**Pull request by cp-Coder**
# Feature Request

## Proposed Changes

- [x] Added fields required in coronasafe/care_fe#3332
- [x] Added fields required in coronasafe/care_fe#3333
- [x] Added fields required in coronasafe/care_fe#3335
- [x] Added fields required in coronasafe/care_fe#3336
- [x] Added fields required in coronasafe/care_fe#3338
- [x] Added fields required in coronasafe/care_fe#3361

## Associated Issue

- Fixes #859
- Related to coronasafe/care_fe#2218
- Related to coronasafe/care_fe#2738

## Associated PR

- Required By: https://github.com/coronasafe/care_fe/pull/3799

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [x] Linting Complete

@coronasafe/code-reviewers

### CODE REVIEW

**Comment by cp-Coder** [comment]
**SWAGGER DOCUMENTATION**
- [x] HEALTH DETAILS
image
- [x] MEDICAL HISTORY
image

File: care/facility/api/serializers/patient_consultation.py [hunk 1]
```diff
@@ -166,9 +195,48 @@ def update(self, instance, validated_data):
...
+
+            medical_history_serializer = MedicalHistorySerializer(
+                medical_history_object,
+                data={
+                    "patient": consultation.patient.external_id.hex,
+                    "consultation": consultation.external_id.hex,
+                    **medical_history_data,
+                },
+            )
+            medical_history_serializer.is_valid(raise_exception=True)
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:234]: Everything needs to be in an atomic block

File: care/facility/api/serializers/patient_consultation.py [hunk 2]
```diff
@@ -166,9 +195,48 @@ def update(self, instance, validated_data):
...
 
+        try:
+            health_details_object = PatientHealthDetails.objects.get(
+                consultation=consultation
+            )
+
+            health_details_serializer = PatientHealthDetailsSerializer(
+                health_details_object,
+                data={
+                    "patient": consultation.patient.external_id.hex,
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:211]: why is this data resupplied ?

File: care/facility/api/serializers/patient_consultation.py [hunk 3]
```diff
@@ -166,9 +195,48 @@ def update(self, instance, validated_data):
...
+            health_details_serializer.is_valid(raise_exception=True)
+            health_details_serializer.save()
+            medical_history_object = MedicalHistory.objects.get(
+                consultation=consultation
+            )
+
+            medical_history_serializer = MedicalHistorySerializer(
+                medical_history_object,
+                data={
+                    "patient": consultation.patient.external_id.hex,
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:229]: This should already be present in the object right?

File: care/facility/api/serializers/patient_consultation.py [hunk 4]
```diff
@@ -166,9 +195,48 @@ def update(self, instance, validated_data):
                 validated_data["kasp_enabled_date"] = localtime(now())
 
         _temp = instance.assigned_to
+        health_details_data = validated_data.pop("new_health_details", None)
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:198]: `None` cannot be unpacked, try something that resembles the actual data's type

File: care/facility/api/serializers/patient_consultation.py [hunk 5]
```diff
@@ -238,14 +306,65 @@ def create(self, validated_data):
...
+            health_details = health_details_serializer.save()
+            consultation.health_details = health_details
+            medical_history_serializer = MedicalHistorySerializer(
+                data={
+                    "patient": consultation.patient.external_id.hex,
+                    "consultation": consultation.external_id.hex,
+                    **medical_history_data,
+                }
+            )
+            medical_history_serializer.is_valid(raise_exception=True)
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:338]: Make sure everything is in an atomic transaction

File: care/facility/api/serializers/patient_consultation.py [hunk 6]
```diff
@@ -238,14 +306,65 @@ def create(self, validated_data):
...
+            medical_history = medical_history_serializer.save()
+            consultation.medical_history = medical_history
+        except KeyError as error:
+            if consultation.patient.last_consultation is None:
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:342]: How would this situation arise?

File: care/facility/models/patient.py [hunk 7]
```diff
@@ -766,3 +761,130 @@ class PatientNotes(FacilityBaseModel, PatientRelatedPermissionMixin):
...
     note = models.TextField(default="", blank=True)
+
+
+class PatientHealthDetails(PatientBaseModel, PatientRelatedPermissionMixin):
+    patient = models.ForeignKey(
+        PatientRegistration,
+        on_delete=models.CASCADE,
+        related_name="health_details",
+    )
+    facility = models.ForeignKey(
```
Comment (vigneshhari) [inline care/facility/models/patient.py:772]: consultation is already 1-1 tied to a facility. consultation is also tied 1-1 to a patient

**Review by vigneshhari — CHANGES_REQUESTED** [review]

**Comment by cp-Coder** [comment]
@vigneshhari I have resolved all the requested changes and the PR is good for testing

**Review by khavinshankar — COMMENTED** [review]
@cp-Coder 

erroring out while migrating, can you check it out
image

**Comment by cp-Coder** [comment]
@khavinshankar The issue has been fixed

**Review by khavinshankar — COMMENTED** [review]
@abhiabhi94 migrations looks good,

@nihal467 can you check it end to end

**Comment by khavinshankar** [comment]
@cp-Coder I'm getting this error, after migration,
image

**Comment by gigincg** [comment]
@khavinshankar If you can add the steps to reproduce as well, @cp-Coder Might be able to debug easier 🚀

**Comment by khavinshankar** [comment]
@cp-Coder To Reproduce the above issue,
Step 1: Clone the repo `git clone git@github.com:coronasafe/care_fe.git`
Step 2: Run the app `sudo make up`
Step 3: Seed dummy data `sudo docker exec care python3 manage.py load_dummy_data`
Step 4: Pull the changes from this pr `gh pr checkout 1073`
Step 5: Migrate the data, `sudo docker exec care python3 manage.py migrate`
Step 6: Try serializing patient instance in django shell `sudo docker exec -it care python3 manage.py shell`
 ```py
from care.facility.models.patient import PatientRegistration
from care.facility.api.serializers import PatientDetailsSerializer
PatientDetailsSerializer(PatientRegistration.objects.first())
 ```
 
 cc: @gigincg

**Comment by khavinshankar** [comment]
@nihal467 had a huddle with @cp-Coder, everything is working fine, and there was no error, I passed in the PatientRegistration model inside PatientConsultationSerializer

@gigincg @vigneshhari can you review this and merge it

**Pull request closed without merging**